blockchain/utxolookup.go

```diff
@@ -0,0 +1,548 @@
+	isCoinBase    bool              // Whether this entry is a coinbase tx.
+	blockHeight   int32             // Height the block containing the tx.
+	sparseOutputs map[uint32]uint32 // Sparse map of unspent output indices.
+	outputs       []utxoOutput      // Packed slice of unspent outputs.
+}
+
+// Version returns the version of the transaction the utxo represents.
+func (entry *UtxoEntry) Version() int32 {
+	return entry.version
+}
+
+// IsCoinBase returns whether or not the transaction the utxo entry represents
+// is a coinbase.
+func (entry *UtxoEntry) IsCoinBase() bool {
+	return entry.isCoinBase
+}
+
+// BlockHeight returns the height of the block containing the transaction the
+// utxo entry represents.
+func (entry *UtxoEntry) BlockHeight() int32 {
+	return entry.blockHeight
+}
+
+// IsOutputSpent returns whether or not the provided output index has been
+// spent based upon the current state of the unspent transaction output store
+// the entry was obtained from.
+//
+// Returns true if the output index references an output that does not exist
+// either due to it being invalid or because the output is not part of the store
+// due to previously being spent/pruned.
+func (entry *UtxoEntry) IsOutputSpent(outputIndex uint32) bool {
+	packedIndex, ok := entry.sparseOutputs[outputIndex]
+	if !ok {
+		return true
+	}
+
+	return entry.outputs[packedIndex].spent
+}
+
+// SpendOutput marks the output at the provided index as spent.  Specifying an
+// output index that does not exist will not have any effect.
+func (entry *UtxoEntry) SpendOutput(outputIndex uint32) {
+	packedIndex, ok := entry.sparseOutputs[outputIndex]
+	if !ok {
+		return
+	}
+
+	// Nothing to do if the output is already spent.
+	output := &entry.outputs[packedIndex]
+	if output.spent {
+		return
+	}
+
+	entry.modified = true
+	output.spent = true
+	return
+}
+
+// IsFullySpent returns whether or not the transaction the utxo entry represents
+// is fully spent.
+func (entry *UtxoEntry) IsFullySpent() bool {
+	// The entry is not fully spent if any of the outputs are unspent.
+	for _, output := range entry.outputs {
+		if !output.spent {
+			return false
+		}
+	}
+
+	return true
+}
+
+// Amount returns the amount of the provided output index.
+//
+// Returns 0 if the output index references an output that does not exist
+// either due to it being invalid or because the output is not part of the store
+// due to previously being spent/pruned.
+func (entry *UtxoEntry) Amount(outputIndex uint32) int64 {
+	packedIndex, ok := entry.sparseOutputs[outputIndex]
+	if !ok {
+		return 0
+	}
+
+	return entry.outputs[packedIndex].amount
+}
+
+// PkScript returns the public key script for the provided output index.
+//
+// Returns nil if the output index references an output that does not exist
+// either due to it being invalid or because the output is not part of the store
+// due to previously being spent/pruned.
+func (entry *UtxoEntry) PkScript(outputIndex uint32) []byte {
```

Comment:
Here's a suggestion for a more descriptive function name instead of `PkScript` (which sounds like an attribute): `PkScriptByIndex`.